dashboard/src/api/auth-manager.ts

```diff
@@ -0,0 +1,138 @@
+import * as axios from 'axios';
+
+import { Api } from './index';
+import { AuthToken, CodeToken } from './types';
+
+class TokenStorage {
+    storeToken(token: string) {
+        if (!token) {
+            throw new Error('Cannot store empty token');
+        }
+        localStorage.setItem('AUTH_TOKEN', token);
+        localStorage.setItem('AUTH_TOKEN_STORED_AT', Date.now().toString());
+    }
+
+    getToken(): string | null {
+        return localStorage.getItem('AUTH_TOKEN');
+    }
+
+    clearToken() {
+        localStorage.removeItem('AUTH_TOKEN');
+        localStorage.removeItem('AUTH_TOKEN_STORED_AT');
+    }
+
+    getTokenAge(): number | null {
+        const timestampStr = localStorage.getItem('AUTH_TOKEN_STORED_AT');
+        if (!this.getToken() || !timestampStr) {
+            return null;
+        }
+        return Date.now() - Number(timestampStr);
+    }
+}
+
+const tokenStorage = new TokenStorage();
+
+export default class AuthManager {
+    /** Maximum age of the token in seconds before it is refreshed. */
+    public maxTokenAge: number = 5 * 60 * 1000;  // 5 minutes
+
+    private _api: Api;
+    private _axios: axios.AxiosInstance;
+    private _authInterceptorId: number | null = null;
+    private _tokenRefreshPromise: Promise<AuthToken> | null = null;
+
+    constructor(api: Api, axiosInstance: axios.AxiosInstance) {
+        this._api = api;
+        this._axios = axiosInstance;
+    }
+
+    checkExistingLogin(): Promise<AuthToken|null> {
+        return this.refreshLogin().then((authToken) => {
+            if (authToken) {
+                this._mountAxiosInterceptor();
+            }
+            return authToken;
+        });
+    }
+
+    initiateLogin(username: string, password: string): Promise<CodeToken> {
+        return this._axios.post<CodeToken>(
+            this._api.endpoints.authCodeToken, {username, password}).then(
+                (response) => response.data);
+    }
+
+    continueLogin(
```

Comment:
Style bikeshedding suggestion:
This function declaration is on two lines but the one before it on one line. You probably have a hard limit on 79 chars so I guess that is the reason but it really hurts readability imho.

Reply:
Long lines hurt the readability more IMO 😄

Reply:
So does inconsistency 😉 But again, bikeshedding, no need for changes.